### ISSUE TITLE
Dirchlet/Value boundary conditions at the top and bottom.

### ISSUE THREAD

**Issue by ali-ramadhan** [climate-machine/Oceananigans.jl#166]
@mukund-gupta has been trying to run some model tests but might need to set Dirchlet/`Value` boundary conditions for the velocities at the top and bottom.

@glwagner would this be easy to add in? For `Flux` boundary conditions I assume we'd have to use 𝜈 instead of κ so something like
 ```julia
Gϕ[i, j, 1] += 𝜈*top_gradient.condition / grid.Δz
```
 instead of
```julia
Gϕ[i, j, 1] += κ*top_gradient.condition / grid.Δz
```
And for `Value` boundary conditions I assume it's just the line below?
```julia
ϕ[i, j, 1] = top_value.condition  # Note we're setting ϕ, not adding to Gϕ.
```

I can try adding these in, shouldn't be too hard but just want to double check.

**Comment by glwagner** [comment]
All that needs to be done is to add a function that computes the flux for the new boundary condition type. 

I’ve designed it so that “diffusivity” is an input to the function that implements the boundary condition. That way you only write one function for a value BC, and it applies to momentum or tracers alike. 

I’ll write more when I get to a computer.

**Comment by glwagner** [comment]
Ok, a couple of important notes about the `apply_bcs!` function:

1. The function takes a "diffusivity" as input. Thus, it is not necessary to have different functions for viscosity and diffusivity. Functionally viscosity *is* diffusivity --- for momentum. See this line:

https://github.com/climate-machine/Oceananigans.jl/blob/4c627e89efe5c268c26d458059bf1904995ebd5b/src/time_steppers.jl#L307

where the viscosity `𝜈` is passed as an input to `apply_bcs`. On this line:

https://github.com/climate-machine/Oceananigans.jl/blob/4c627e89efe5c268c26d458059bf1904995ebd5b/src/time_steppers.jl#L296

`𝜈` is determined to be `cfg.𝜈v`. This will change in the future when we have closures that can modify the viscosity/diffusivity (and hopefully the default viscosity and diffusivity will be made isotropic and defined as fluid 'constants', like gravitational acceleration). We also will need to modify `apply_bcs!` when the viscosity at top and bottom is different.

2. To add a new boundary condition (in the current set-up with periodic lateral boundaries), one needs to define `apply_z_top_bc!` and `apply_z_bottom_bc` for that boundary condition. See this line and below in `boundary_conditions.jl`:

https://github.com/climate-machine/Oceananigans.jl/blob/4c627e89efe5c268c26d458059bf1904995ebd5b/src/boundary_conditions.jl#L155

Notice that `ϕ` is an arbitrary field, and `κ` is the diffusivity/viscosity for `ϕ`.

Let me know if this makes sense. If it doesn't, we should think about ways we can simplify the boundary condition implementation (and I am happy to implement `Value` boundary conditions.

**Comment by glwagner** [comment]
Just read the original post and I realized there's an additional point to clarify:

3. `Flux` and `Gradient` boundary conditions *are* currently implemented for momentum. In other words, this line:

https://github.com/climate-machine/Oceananigans.jl/blob/4c627e89efe5c268c26d458059bf1904995ebd5b/src/time_steppers.jl#L307

is called for both `u` and `v`. Go ahead and try it!

**Comment by ali-ramadhan** [comment]
Thanks @glwagner! This makes a lot of sense. @mukund-gupta and I implemented a `NoSlip` boundary condition type.

We're just figuring out how to impose `Value` boundary conditions correctly on _u_ and _v_ at the bottom. Have you thought about this? Might be complicated because of the non-hydrostatic pressure corection.

cc @christophernhill @jm-c

**Comment by glwagner** [comment]
Okay!

Note `NoSlip` is not different from a `Value` condition with the value set to 0. So there's a redundancy if you have `NoSlip` --- two apparently different boundary conditions are actually identical.

In `OceanTurb.jl`, I interpret the need for a 'no-slip' condition as a user API requirement, rather than an internal code requirement. Thus I implement a special API function `ZeroFluxBoundaryConditions`, rather than creating a new type for a flux condition when the flux is zero.

I'll think about the pressure solver. Are you saying that the pressure boundary conditions are no longer homogeneous Neumann (what the pressure solve assumes) with a non-zero flux at the bottom?

Note that this wouldn't be unique to `Value` boundary conditions; it should apply to all inhomogeneous boundary conditions (with respect to the flux of a quantity) if so.

**Comment by ali-ramadhan** [comment]
Ah you're right, that would make sure sense. We can change our `NoSlip` to actually do `Value` boundary conditions. Not sure how to impose a non-zero value but once we figure this out we can change them.

**Comment by glwagner** [comment]
@ali-ramadhan imposing a non-zero value is identical to how a value of zero is imposed: the flux divergence of the quantity is calculated based on the gradient between the value in the cell above the wall, and the imposed value on the wall, and added to the RHS.

See how it's done in `OceanTurb.jl`:

https://github.com/glwagner/OceanTurb.jl/blob/990be4f4fcf0c55a4a5c31582d282b58e6f40935/src/boundary_conditions.jl#L158

**Comment by ali-ramadhan** [comment]
Ah I see, thanks for the explanation! Does this kind of look correct then?
```julia
@inline apply_z_bottom_bc!(bottom_value::BC{<:Value, <:Number}, ϕ, Gϕ, ν, t, grid, u, v, w, T, S, iteration, i, j) =
    Gϕ[i, j, grid.Nz] -= (2/grid.Δz^2) * ν * (ϕ[i, j, grid.Nz] - bottom_value)
```

**Comment by glwagner** [comment]
The numbers look right (I'm not sure about the sign). This won't work for variable grid spacing though?

In the other boundary condition functions the diffusivity is called `κ`. Might be best to use `κ` here for consistency.

What I was going to say:

The flux divergence contribution from the wall is `flux / Δc`, and the `flux` at the bottom (for example) is

```
flux = -2 * κ * (c_bottom_cell - c_bndry) / Δf
```

Note that we define flux so that the contribution to the RHS is minus the flux divergence.

**Comment by glwagner** [comment]
Ok, so I guess due to the reverse indexing convention, the quantity

```julia
ϕ[i, j, grid.Nz] - bottom_value
```

is the negative of the gradient of `ϕ`?

Therefore the flux is 

```julia
flux = 2 / grid.Δz * ν * (ϕ[i, j, grid.Nz] - bottom_value)
```

and minus the flux divergence associated with the wall flux is 

```julia
-2 / grid.Δz^2 * ν * (ϕ[i, j, grid.Nz] - bottom_value)
```

**Comment by ali-ramadhan** [comment]
That makes a lot of sense!

Yes I think the minus sign comes from the difference in indexing... We're still with `k=1` at the top surface. We tested it with `bottom_value = 0` and the velocity was spinning down under friction at the bottom so we must have been doing something right.

We can implement the `Value` boundary conditions. There might be an analytic solution for the spin-down of a fluid with constant velocity under friction that we can test against.

**Comment by glwagner** [comment]
Ok, I see where I made my mistake. The quantity

```julia
ϕ[i, j, grid.Nz] - bottom_value
```

is actually *proportional* to the gradient of `ϕ` (not the negative of the gradient of `ϕ`).

There are many solutions. Couette flow and Poiseuille flow are classic cases. The diffusive relaxation of the initial value problem with a linear gradient + cosines can also be solved analytically for u, v, T, or S.

**Comment by ali-ramadhan** [comment]
@glwagner was this resolved in #237?
